### ISSUE TITLE
Naming inline modules

### ISSUE THREAD

**Issue by timbur** [whatwg/loader#42]
I was referred here to discuss this.  Apologies if I'm going about it wrong.  :)

The idea is to allow inline modules to be cached within the loader just like any other module by giving `<module>` a `name` attribute.

Suppose you have some inline module within HTML like so:

```html
<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv='Content-type' content='text/html; charset=utf-8'>
    <title>Example</title>
  </head>
  <body>
    <module name="main">

        import App from 'some/dynamic/location';
        var ImmediateBlob = {"maybe": "some JSON stringified by server upon request"};

        App.doSomething();
        export { App, ImmediateBlob };

    </module>
    <script src="some/static/script.js"></script>
  </body>
</html>
```

And `some/static/script.js` might look for the `main` module:
```js
import main from 'main';

main.App.looksLikeADuck();

if (main.ImmediateBlob.quacks == 'like a duck') {
    console.log('It is a duck!');
}
```

Two other relevant discussions:
https://github.com/ModuleLoader/es6-module-loader/pull/343
https://github.com/jorendorff/js-loaders/issues/83#issuecomment-33991024

**Comment by matthewp** [comment]
This would not be possible under the current spec due to there not being a concept of module names any more.  Everything is keyed by url so virtual modules are an awkward fit now (hopefully this decision will be reconsidered, read #28). Add this to the pile of benefits of module ids over urls.

Also, neither System.define nor System.module have counterparts in the current spec.

**Comment by timbur** [comment]
Surely it makes the most sense, as @dherman pointed out at https://github.com/jorendorff/js-loaders/issues/83#issuecomment-33991024, to do the following:

1. Associate inline modules **without** a `name` attribute with `System.module`.
2. Associate inline modules **with** a `name` attribute with `System.define`, but only if a path is not set for that `name`.  (This should have the added advantage of showing up within the sources panel as something other than `<Anonymous Module>`.)
3. For any modules with a `src`, use the **address-normalized** path as usual.  And for any modules with a `name`, use the **name-normalized** path as long as it does not yet exist.  It should also be possible to have both the `name` and the `src` paths.

I can't imagine any conflicts occurring with this system.

**Comment by caridy** [comment]
@matthewp that's not accurate, you can have any arbitrary entry in the registry.

`Loader.install()` will allow you to define new entries in the registry: https://whatwg.github.io/loader/#reflect-loader-install

The normalization process (to a final url), including the `sites()` configuration will happen if there is no entry available in the registry for arbitrary values in the module specifier.

**Comment by matthewp** [comment]
@caridy Interesting, `install()` doesn't specify a return value, this means it could be a Promise.  This differs from the old System.set / System.get.

**Comment by matthewp** [comment]
@timbur This spec is quite a bit different from the one referenced in that thread, we need a `Loader.define` and `Loader.module` first.  I'm not sure what the authors' opinion is on `.define`, I remember there was some problems with the old version.  `.module` will almost definitely be needed to support `<module>`.  Maybe a separate issue to track that?

**Comment by timbur** [comment]
In my opinion, https://github.com/ModuleLoader/es6-module-loader and https://github.com/systemjs/systemjs is handling everything perfectly and intuitively, so it would make sense to model anything missing from this spec after their functionality.  Plus, those projects are designed to match this spec, so it's a bit of a win-win for everyone.

**Comment by caridy** [comment]
@timbur systemjs should not dictate what we do here, it is the other way around. We are considering those ideas that we have tried, and play around with in systemjs, but ultimately we have to be pragmatic.

@matthewp I think you're confused. `install()` does not fetch deps since you have to provide a [[Module]] instance, which produces a new entry in the registry with `[[State]]: "ready"`, there is nothing to return.

**Comment by timbur** [comment]
I didn't mean to imply that SystemJS should dictate anything here.  I was only saying that there is some extra functionality present within SystemJS which makes a lot of sense, and this functionality exists because of practical, real-world necessity/usefulness.  I'll provide an example of said necessity shortly (in addition to the example I provided above).

**Comment by guybedford** [comment]
This spec has a `loader.provide` which actually gives more flexible behaviour than the `System.define`, `System.module`. See https://github.com/ModuleLoader/es6-module-loader/blob/1.0/src/loader.js#L429 for how it works.

@timbur note that `name` is deprecated in this spec, and modules are only `src` based.

**Comment by matthewp** [comment]
@caridy I might be confused, let's break this down:

> The normalization process (to a url), including the sites() configuration will happen if there is no entry available in the registry for an arbitrary values in the module specifier.

This normalization process is (potentially) asynchronous, if it goes through the `resolve` hook.  If that's the case it **should** return something.

**Comment by caridy** [comment]
@matthewp the normalization process only happens when importing a module (or a dep), not when installing a new entry in the registry.

**Comment by matthewp** [comment]
Ok, then I did misunderstand you.  So this mean keys are really just arbitrary strings, I can do:

```js
System.install("@@foobar", { default: { function() { return "bar"; } });
```

And anyone importing that will get the correct module, cool.

**Comment by timbur** [comment]
@guybedford Can that same flexible behavior be achieved directly through `System`?  Or is it currently limited to methods within `loader`?

As for the real-world necessity, it makes no sense to render inline modules inaccessible.  They're modules, too, right?  It should be possible to import them.  It also doesn't make sense to give them the `src` attribute since the source is contained within itself.  The only logical conclusion, in my opinion, is to use `name` so that said inline modules can be imported by their names.

The way I see it, from a simple point of view and for predictability's sake, internally, when paths are set and modules are cached, all that matters in the end is the corresponding key.  A few simple rules and everything gets insanely simple:
1. Is there a path match?  If so, use it and adjust the key to its full path name.
2. Are there any identifiers that would indicate it's a file (`src`, has a file extension, etc.)?  If so, use it and adjust the key to its full path name.
3. A single word with no extension?  Don't do anything fancy.  Don't change it.  Just use it.  At this point in these 3 steps, it's safe to assume we're looking for an inline module that was cached using `<module name="...">` or explicitly defined some other way.  And if it isn't already cached, hit the server without appending `.js` to it.

Anything beyond that would be convoluted, adding unnecessary complexity and wasting everyone's time.  Keep it as simple and predictable as that and the only conceivable way developers could screw up is if they purposely shoot themselves in the foot.

**Comment by MajorBreakfast** [comment]
I like the simplicity of having just `src`.

What about this:
```JS
<module src="hello.js">
export default 'I am a module'
</module>
```

**Comment by matthewp** [comment]
@MajorBreakfast that means `src` has a different meaning depending on whether there is inner text or not.  Would also different from every other use of `src` attributes in html.

**Comment by MajorBreakfast** [comment]
@matthewp Hmm, yes. Maybe `<script>` tag + `System.install()` comes close enough already. Likely, there even should not be a more convenient way to do this.

**Comment by caridy** [comment]
@timbur no. we have discarded that a while ago, and we've settled on the idea that inline modules will probably be focus on initialization (just like inline scripts are today), and therefore there is really no need to export things into the runtime from them, instead they just import other modules, and carry on with some initialization routines. I'm curious about what sort of inline script code you have today that modifies the runtime to extend the available functionality.

**Comment by matthewp** [comment]
@guybedford I missed your comment but you're right that `.provide` is a better `.define`. Don't we still need a key-less form, though?  By default a `<module>` shouldn't be part of the registry.

**Comment by timbur** [comment]
@caridy My issue with that line of thinking is that if developers want to do something, they'll figure out a way to do it.  It's always better to provide an elegant, predictable solution for people to use, rather than imposing arbitrary restrictions that will lead developers to create ugly hacks/workarounds.  And for this particular case I'm 100% certain that will happen.

Consider this scenario:  Users load some module specific to their profile.  We'll refer to it as a `duck`.  Every user has their own `duck` with custom functionality that could not be easily defined in any form other than JavaScript.  It's initialized upon loading the page, and many other modules want to use that particular `duck`, expecting a `quack` method that is potentially custom (different) for each user.

Suppose we have my duck at `timbur/duck.js`:
```js
export function quack () {
    // let's pretend this is all kinds of complex logic
    console.log('quack!');
}
```

And we have your duck at `caridy/duck.js`:
```js
export function quack () {
    // let's pretend this is all kinds of complex logic
    alert('quack!');
}
```

And when I load the page, the HTML generated for me looks something like this:
```html
<module>
    import duck from 'timbur/duck.js';
    duck.quack();
</module>
```

And yours of course looks something like this:
```html
<module>
    import duck from 'caridy/duck.js';
    duck.quack();
</module>
```

And later on, perhaps some other module wants to make the user's particular `duck` `quack`, we'll call it `goose.js`.  There are of course many different ways to make this happen, so let's consider a few.

Without the ability to name (and register) the `<module>` above, it's really not even a module.  It's just a script.  So, as a script, if we want to maintain some reference to the user's `duck`, the script itself will need to assign the `duck` to some global variable.
```html
<script>
    import duck from 'caridy/duck.js';
    duck.quack();
    window.main = {duck: duck};
</script>
```

Or for the sake of consistency and "elegance" (if you could call it that), one might pull some other module designed specifically for storing this information:
```html
<script>
    import store from 'store.js';
    import duck from 'caridy/duck.js';
    duck.quack();
    store.set('duck', duck);
</script>
```

And then for our `goose.js` module to use the user's particular `duck`:
```js
window.main.duck.quack();
```

Or:
```js
import store from 'store.js';
store.get('duck').quack();
```

Now let's be honest.  The above solutions are ugly and they go against the grain of the module system.

Of course, all of this so far might seem like a contrived example, and most people might find this particular usage an edge case.  But again, it's never a good idea to impose restrictions on what should or shouldn't be possible only because it isn't some scenario that is easy to imagine.  People are obviously already aware of these possibilities, so why limit them?  It's much better to provide an elegant solution at the core.

And finally, let's consider a solution that I would probably use if naming an inline module was **not** built into the core of `<module>`'s functionality.  I realize this solution relies on `System.define`, but let's not be pedantic.  The analogy remains.

HTML snippet:
```html
<module name="main">
    import duck from 'timbur/duck.js';
    duck.quack();
    export duck;
</module>
<script src="register-inline-modules.js"></script>
```

`register-inline-modules.js`:
```js
var modules = document.getElementsByTagName('module');

for (var i = 0; i < modules.length; i++) {

    var module = modules[i];

    if (module.hasAttribute('name')) {
        var name = module.getAttribute('name');
        var source = module.innerHTML.substr(1);
        System.define(name, source);
    }

}
```

Then `goose.js` would very predictably look like this:
```js
import main from 'main';
main.duck.quack();
```

That is way simpler and follows the module paradigm much more consistently.  But we aren't finished yet.

Last but certainly not least, consider that the `name` attribute is supported.  The following is all we need.  No global pollution.  No modules to designed specifically to remember `duck`.  No unnecessary extra scripts.  It's intuitive and succinct.  Just use the registry.

HTML snippet:
```html
<module name="main">
    import duck from 'timbur/duck.js';
    duck.quack();
    export duck;
</module>
```

`goose.js`:
```js
import main from 'main';
main.duck.quack();
```


It's entirely possible that I'm being stupid and completely overlooking an easier way to achieve this.  And if that is the case, please let me know and I will gladly shut up about this particular example.  Either way, I think there are more than enough use cases out there that we haven't thought of which will resemble the above functionality, so why not make it elegant?

**Comment by MajorBreakfast** [comment]
@timbur I think the idea is to have a `main.js` file (Directly or indirectly through the use of a bundling mechanism like in SystemJS). The module tag then simply imports that module.

> It's always better to provide an elegant, predictable solution for people to use [...]

Not if the solution can be considered a dead end.

**Comment by timbur** [comment]
But it isn't a dead end.  The key point is that it should be possible to elegantly register dynamic modules under some common name without resorting to extra scripts.  Yes, in production, all static modules will be minified and bundled.  But it would be stupid to create separate bundles for each user due to some dynamic module reference.  The dynamic module would obviously automatically be excluded from the bundle, and the most efficient way for modules within the bundle to use it is to have its reference exported upon initialization.  I don't know how I can make it any clearer, and I've yet to see a reasonable argument against this functionality along, nor a more elegant solution to the problem.

**Comment by johnjbarton** [comment]
Can the same effect be created with

<script>
System.sites = {main: 'timbur/main.js'};
</script>

?
Like Tim's example, this would be a per user specific page script. Like
Tim's example, the effect is to cause goose.js to see a different main
module dynamically, when it runs import. Like Tim's example it makes
development more challenging because the module is dynamic.

Unlike Tim's example it does not require a <module> with a name.

(To be sure, IMO we don't need a <module> tag without a name either. A
<script> tag with System calls are clearer and provide better developer
ergonomics in realistic code.)

jjb

**Comment by timbur** [comment]
That's a pretty solid suggestion @johnjbarton, albeit still not nearly as elegant and intuitive as it could be.  And I actually agree that a `<script>` tag with `System` calls is more than sufficient.  However, I personally feel that web development in the future will take place almost entirely within the browser itself.  I'll explain my reasoning and what it has to do with modules.

We're already seeing a ton of movement towards minimizing the feedback loop with development - e.g., webpack hot module replacement, react-hot-loader, workspaces within dev tools that automatically save changes to CSS within the inspector, etc., etc.  So how can we get the feedback loop to an absolute minimum?  By having all of our development tools directly in the browser.  And of course, modules are undoubtedly the future, so we need a really solid module system that works seamlessly with this paradigm.  It needs to be **effortless** to develop modules and make them work together, and even inline modules should be identifiable.  It's important to know what's what.

With that said, let's compare what an in-browser development environment would look like without named modules.  Also consider that there might end up being more than a handful of anonymous modules in the development environment, for whatever purposes.
![anonymous-module](https://cloud.githubusercontent.com/assets/7020411/6986464/6493fdfc-da0b-11e4-8b5c-8ce2cae6cfef.gif)

And this is what it looks like with the named inline module.  It's more intuitive, much easier to decipher.
![named-module](https://cloud.githubusercontent.com/assets/7020411/6986572/791fc05c-da0c-11e4-9e50-bd59f2025ad2.gif)

**Comment by caridy** [comment]
@timbur are you talking about bucket testing? jejejeje

Anyway, @johnjbarton is righty, `loader.sites()` is really the way to go for those use-cases.

**Comment by MajorBreakfast** [comment]
@timbur Each HTML file should only contain one module tag. The named file that shows up in the dev tools cannot be edited because it exists on disk inside the index.html. **Why not use a `main.js` file?**

**Comment by timbur** [comment]
@caridy I'm not talking about bucket testing at all, but that's a perfect example of one of the many possibilities that would be easier (or more elegant, at least) if naming inline modules is supported.  Can you provide an example using `loader.sites()`?

@MajorBreakfast You're thinking inside the box.  In an ideal in-browser development workflow, there are all kinds of possibilities that could potentially result in multiple html files and thus, multiple inline modules.  As for using a `main.js` file, the code would be different for every user, so it makes much more sense to be an inline script.

**Comment by caridy** [comment]
@timbur and co, our goal is not to provide sugar for every use case but for the most common ones, and make sure that the rest can be implemented in userland. This use case can be implemented in userland, no doubt about it, and We haven't seen this one very often when people embrace the ES module philosophy, It is really more like a side effect of trying to make modules to behave like scripts, which is fine but ... it is not really something we will promote or try to provide sugar for. If you really want to continue this conversation, you should demontrate that the currect spec does not allow you to do it on userland, or came up with a real very common use case that will requiere named inline modules. Makes sense?

**Comment by timbur** [comment]
Of course just about anything can be achieved in userland.  The point I've made is that including this functionality within the spec would only result in one or two lines extra of code at core of `<module>`, and these one or two lines would result in more than enough predicted benefits to warrant its inclusion, along with possibly many many more unpredicted benefits and no downsides whatsoever.

**Comment by probins** [comment]
> came up with a real very common use case that will requiere named inline modules

one that I've just been investigating is HTML Imports. Currently most existing web components that I've seen that use HTML Import have an inline template plus an inline script to deal with the template. In some cases, that script will just be an initialisation of some sort, but in others the component can be interactive and expose some sort of API - getThis, setThat. If these scripts are converted to modules, either there will have to be a rule that they should always link to an outside file (which might make sense from CSP point of view), or they will have to be named, so other modules can import them.

**Comment by probins** [comment]
a simple example of this is Polymer. https://github.com/Polymer/polymer/blob/master/dist/polymer.html is the main component which runs a script to load the Polymer class function which all Polymer components use. This is currently a global. If you change this structure to modules, then all components have to import the Polymer module. To do that, it has to be named.

**Comment by MajorBreakfast** [comment]
Custom elements are also global.

**Comment by guybedford** [comment]
One could imagine that script being upgraded to a module:

```html
<script type="module" src="polymer.js"></script>
```

Where 'src' above is an unnormalized module name (it may be called 'key' or something else though).

There is no need to name the module in the above, it is loaded anonymously and stored in the registry like any normal module load. Defining how modules affect rendering is the next question there though.

**Comment by probins** [comment]
@guybedford You're saying that `<module src="x.js">` would be handled differently from `<script src="x.js">`? See the comments from Apr 2. And what about inline modules in HTML Imports? If they export values, they need to have an identifier.

The basic issue here is that you are not importing a module file with a url, you are importing a document fragment which contains one or more module definitions which may need to be loaded in the registry, System.define() in the old spec, loader.install() in the new.

**Comment by probins** [comment]
@MajorBreakfast HTML imports don't have to be custom elements. They could be a form or a bunch of divs or whatever grouping of html elements you like.

**Comment by probins** [comment]
Having raised this issue, I'm not sure there is actually a problem. :-) The simple solution for the case where a component has to export values is to split the code in two: the `<module>`, whether inline or external, does the initial setting up, and if there's a need for exporting, then that code can be put in a separate js file which can be imported in the usual way. A disadvantage of this would be that the separate js file would be detached from the HTML Import and not necessarily linked to from it. But this is also the case with the main document: you can't tell from the html which modules are loaded (form part of the app). In the Polymer case, I don't think you need a Polymer component with a `<module>`, just a standard module for the Polymer class that can be imported by Polymer components.

In summary, a web app is an html document that can import stylesheets and components, and can contain an anonymous `<module>` to bootstrap the app. A component is an html document fragment that can ... to bootstrap the component.

**Comment by caridy** [comment]
correct, none of these examples really require an inline module to be named, lets keep it that way for now. We can always revisit this if needed :)
